### PR TITLE
docs: Add docs about schema structs

### PR DESCRIPTION
**Pull request by evenyag**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Add docs about the following schema structs
- ColumnSchema
- Schema
- RawSchema
- RegionSchema
- StoreSchema
- ProjectedSchema

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
- #741

### CODE REVIEW

File: docs/schema-structs.md [hunk 1]
```diff
@@ -0,0 +1,438 @@
...
+    pub schema: SchemaRef,
+    pub primary_key_indices: Vec<usize>,
+    pub value_indices: Vec<usize>,
+    // ...
+}
+```
+
+The order of columns in `TableMeta::schema` is the same as the order specified in the `CREATE TABLE` statement which users use to create this table.
+
+`TableMeta::schema` doesn't have metadata about which columns are primary key columns or value columns (non-primary key and time index). We store them in `primary_key_indices` and `value_indices`.
```
Comment (WenyXu) [inline docs/schema-structs.md:30]: Is this imply the `TableMeta::schema` is an `arrow_schema::SchemaRef` type? In this context, we have two kinds of `Schema`, which is confusing.

File: docs/schema-structs.md [hunk 2]
```diff
@@ -0,0 +1,438 @@
...
+    ts TIMESTAMP,
+    host STRING,
+    usage_user DOUBLE,
+    usage_system DOUBLE,
+    datacenter STRING,
+    TIME INDEX (ts),
+    PRIMARY KEY(datacenter, host)) ENGINE=mito WITH(regions=1);
+```
+
+Then the table's `TableMeta` may look like this:
```
Comment (WenyXu) [inline docs/schema-structs.md:44]: Something like `RawTableMeta` (like `RawSchema`)? 
As mentioned above:
> `TableMeta::schema` doesn't have metadata about which columns are primary key columns or value columns (non-primary key and time index). 

However, the serialized schema shows it contains `time_index`.

File: docs/schema-structs.md [hunk 3]
```diff
@@ -0,0 +1,438 @@
...
+pub struct RegionSchema {
+    user_schema: SchemaRef,
+    store_schema: StoreSchemaRef,
+    columns: ColumnsMetadataRef,
+}
+```
+
+Each region reserves some columns called `internal columns` for internal usage:
+- `__sequence`
+- `__op_type`
```
Comment (WenyXu) [inline docs/schema-structs.md:96]: What is the `__op_type` column?

File: docs/schema-structs.md [hunk 4]
```diff
@@ -0,0 +1,438 @@
...
+        ],
+        "time_index":2,
+        "version":0
+    },
+    "row_key_end":3,
+    "user_column_end":5
+}
+```
+
+The key and timestamp columns form row keys of rows. We put them together so we can use `row_key_end` to get a slice of all row key columns. Similarly, we can use the `user_column_end` to get a slice of all user columns (non-internal columns).
```
Comment (WenyXu) [inline docs/schema-structs.md:174]: Maybe add a simple example like `row identifier := columns(0..row_key_end) `?

**Review by WenyXu — COMMENTED** [review]
Hurray 🥳

File: docs/schema-structs.md (hunk 1 above)
Comment (evenyag) [reply docs/schema-structs.md]: It's our `Schema` type. I didn't mention `arrow::SchemaRef`. I put our `SchemaRef` definition in the doc to make it more obvious.

File: docs/schema-structs.md (hunk 3 above)
Comment (evenyag) [reply docs/schema-structs.md]: This is out of this document's topic, so I only add a brief description.

**Comment by v0y4g3r** [comment]
The ***Master of Schemas***

**Comment by v0y4g3r** [comment]
I think we still need a graph demonstrating conversions between schemas 🧎‍♂️

**Comment by waynexia** [comment]
> I think we still need a graph demonstrating conversions between schemas 🧎‍♂️

Can't agree more. It always feels like falling into a maze when I want to figure out how to retrieve some information🧎‍♂️

**Comment by evenyag** [comment]
I don't want to draw or touch such a graph anymore https://github.com/GreptimeTeam/greptimedb/pull/1373/commits/012c4d52b3fc30afe67068c3c1f9a6d9bd1b5b06. @v0y4g3r  It must deserve $1M 💰.

**Comment by waynexia** [comment]
Genius maze designer https://github.com/GreptimeTeam/greptimedb/commit/012c4d52b3fc30afe67068c3c1f9a6d9bd1b5b06 👑 !

**Comment by v0y4g3r** [comment]
> I don't want to draw or touch such a graph anymore [012c4d5](https://github.com/GreptimeTeam/greptimedb/commit/012c4d52b3fc30afe67068c3c1f9a6d9bd1b5b06). @v0y4g3r It must deserve $1M 💰.

81518dfca48141a69a0338d37855d15d

**Review by v0y4g3r — APPROVED** [review]
LGTM

File: docs/schema-structs.md [hunk 5]
```diff
@@ -0,0 +1,527 @@
...
+        "column_schemas":[
+            "ts",
+            "usage_system"
+        ],
+        "time_index":0
+    }
+}
+```
+
+As you can see, `schema_to_read` doesn't contain column `usage_user`.
```
Comment (MichaelScofield) [inline docs/schema-structs.md:305]:
```suggestion
As you can see, `schema_to_read` doesn't contain the column `usage_user` that is not intended to be read (not in projection).
```

File: docs/schema-structs.md [hunk 6]
```diff
@@ -0,0 +1,527 @@
...
+#[derive(Debug)]
+pub struct ReadAdapter {
+    source_schema: StoreSchemaRef,
+    dest_schema: ProjectedSchemaRef,
+    indices_in_result: Vec<Option<usize>>,
+    is_source_needed: Vec<bool>,
+}
+```
+
+For each column required by `dest_schema`, `indices_in_result` stores the index of that column in the row read from the source memtable or SST. If the source row doesn't contains that column, the index is None.
```
Comment (MichaelScofield) [inline docs/schema-structs.md:323]:
```suggestion
For each column required by `dest_schema`, `indices_in_result` stores the index of that column in the row read from the source memtable or SST. If the source row doesn't contain that column, the index is `None`.
```

**Review by MichaelScofield — APPROVED** [review]
I recommend link all the codes to our cargo doc ("https://greptimedb.rs/greptime/") instead of Github files.

File: docs/schema-structs.md [hunk 7]
```diff
@@ -0,0 +1,527 @@
...
+    pub data_type: ConcreteDataType,
+    is_nullable: bool,
+    is_time_index: bool,
+    default_constraint: Option<ColumnDefaultConstraint>,
+    metadata: Metadata,
+}
+```
+
+## Schema
+[Schema](https://github.com/GreptimeTeam/greptimedb/blob/9fa871a3fad07f583dc1863a509414da393747f8/src/datatypes/src/schema.rs#L38) is an ordered sequence of `ColumnSchema`. It is equivalent to arrow's [Schema](https://docs.rs/arrow/latest/arrow/datatypes/struct.Schema.html) with additional metadata including the index of the time index column and the version of this schema. Same as `ColumnSchema`, we can convert our `Schema` between arrow's `Schema`.
```
Comment (evenyag) [inline docs/schema-structs.md:21]:
```suggestion
[Schema](https://github.com/GreptimeTeam/greptimedb/blob/9fa871a3fad07f583dc1863a509414da393747f8/src/datatypes/src/schema.rs#L38) is an ordered sequence of `ColumnSchema`. It is equivalent to arrow's [Schema](https://docs.rs/arrow/latest/arrow/datatypes/struct.Schema.html) with additional metadata including the index of the time index column and the version of this schema. Same as `ColumnSchema`, we can convert our `Schema` from/to arrow's `Schema`.
```

**Pull request merged**